core/src/main/java/io/substrait/relation/RelCopyOnWriteVisitor.java

```diff
@@ -0,0 +1,445 @@
+    for (int i = 0; i < expressionCount; i++) {
+      Expression oldExpr = oldExpressions.get(i);
+      Expression newExpr = visitExpression(oldExpr);
+      if (oldExpr != newExpr) {
+        if (oldExpressions == newExpressions) {
+          newExpressions = new CopyOnWriteArrayList<>(oldExpressions);
+        }
+        newExpressions.set(i, newExpr);
+      }
+    }
+    return newExpressions;
+  }
+
+  @Override
+  public Project visit(Project project) {
+    Rel oldInput = project.getInput();
+    Rel newInput = oldInput.accept(this);
+    List<Expression> oldExpressions = project.getExpressions();
+    List<Expression> newExpressions = copyOnWrite(oldExpressions);
+    return oldInput == newInput && oldExpressions == newExpressions
+        ? project
+        : ImmutableProject.builder()
+            .from(project)
+            .input(newInput)
+            .expressions(newExpressions)
+            .build();
+  }
+
+  @Override
+  public Sort visit(Sort sort) {
+    Rel oldInput = sort.getInput();
+    Rel newInput = oldInput.accept(this);
+    return oldInput == newInput ? sort : ImmutableSort.builder().from(sort).input(newInput).build();
+  }
+
+  @Override
+  public VirtualTableScan visit(VirtualTableScan virtualTableScan) {
+    return virtualTableScan;
+  }
+
+  private Expression visitExpression(Expression expression) {
+    ExpressionVisitor<Expression, RuntimeException> visitor =
+        new ExpressionVisitor<Expression, RuntimeException>() {
+
+          public interface Getter<T> {
+            public Expression get(T element);
+          }
+
+          public interface Builder<T> {
+            public T build(T element, Expression e);
+          }
+
+          private static <T> List<T> copyOnWrite(
+              ExpressionVisitor visitor, List<T> oldExpressions) {
+            return copyOnWrite(
+                visitor, oldExpressions, element -> (Expression) element, (element, e) -> (T) e);
+          }
+
+          private static <T> List<T> copyOnWrite(
+              ExpressionVisitor visitor,
+              List<T> oldExpressions,
+              Getter<T> getter,
+              Builder<T> builder) {
+            int expressionCount = oldExpressions.size();
+            List<T> newExpressions = oldExpressions;
+            for (int i = 0; i < expressionCount; i++) {
+              T element = oldExpressions.get(i);
+              Expression oldExpr = getter.get(element);
+              Expression newExpr = (Expression) oldExpr.accept(visitor);
+              if (oldExpr != newExpr) {
+                if (oldExpressions == newExpressions) {
+                  newExpressions = new CopyOnWriteArrayList<>(oldExpressions);
+                }
+                newExpressions.set(i, builder.build(element, newExpr));
+              }
+            }
+            return newExpressions;
+          }
+
+          @Override
+          public Expression.NullLiteral visit(Expression.NullLiteral expr) {
```

Comment:
Rather than having this duplicated here, I suggest we create a new class that is AbstractExpressionVisitor that has all visit methods delegate to
```
abstract RETURN visitFallback(Expression expression) throws EXCEPTION;
```

Then we can use that abstract class here and write the fallback behavior as:

```
Optional<Expression> visitFallback(Expression expression) {
  return Optional.empty();
}
```

That way we encapsulate the concept of default/fallback elsewhere and this class can shrink to just the meaty functionality. We don't have all these `return expr;` methods here. (I'm guessing we'll have the fallback behavior elsewhere as well.)

Reply:
Do you mean instead of adding `return Optional.empty()` instead call `visitFallback(expr)` everywhere?

Reply:
I'm saying that the new AbstractExpressionVisitor has all methods call visitFallback. Then this implementation implements visitFallback to return Optional.empty().
```

public abstract class AbstractExpressionVisitor<RETURN, EXCEP extends Exception> implements ExpressionVisitor<RETURN, EXCEP> {

  public abstract RETURN visitFallback(Expression expr);

  public RETURN visit(Expression.NullLiteral expr)  throws EXCEP {
    return visitFallback(expr);
  } 

  public RETURN visit(Expression.BoolLiteral expr)  throws EXCEP {
    return visitFallback(expr);
  } 

  // all the methods calling visitFallback...

}


// in copyonwrite...
ExpressionVisitor<Expression, RuntimeException> visitor =
        new AbstractExpressionVisitor<Expression, RuntimeException>() {

  @Override
  public Expression visitFallback(Expression expression) {
    return Optional.empty();
  }

};
```